pybamm/meshes/meshes.py

```diff
@@ -35,14 +35,33 @@ class Mesh(dict):
 
     """
 
-    def __init__(self, geometry, submesh_types, submesh_pts):
+    def __init__(self, geometry, submesh_types, var_pts):
         super().__init__()
+        # convert var_pts to an id dict
+        var_id_pts = {}
+        for var, pts in var_pts.items():
+            var_id_pts[var.id] = pts
+
+        # create submesh_pts from var_pts
+        submesh_pts = {}
+        for domain in list(geometry.keys()):
+            submesh_pts[domain] = {}
+            for prim_sec in list(geometry[domain].keys()):
+                for var in list(geometry[domain][prim_sec].keys()):
+                    if var.id not in var_id_pts.keys():
+                        raise KeyError(
+                            "Points not given for a variable in domain {}".format(
+                                domain
+                            )
+                        )
+                    submesh_pts[domain][var.id] = var_id_pts[var.id]
         self.submesh_pts = submesh_pts
+
         for domain in geometry:
             repeats = 1
             if "secondary" in geometry[domain].keys():
                 for var in geometry[domain]["secondary"].keys():
-                    repeats = submesh_pts[domain][var.name]  # note (specific to FV)
+                    repeats = submesh_pts[domain][var.id]  # note (specific to FV)
```

Comment:
would be slightly clearer to have 
```python3
else:
   repeats = 1
```
instead of creating `repeats` in line 61 and then maybe overwriting it